plugin/pkg/replacer/replacer.go

```diff
@@ -13,123 +13,134 @@ import (
-		if rr.Msg != nil {
-			rep.replacements[headerReplacer+"rflags}"] = flagsToString(rr.Msg.MsgHdr)
+		return rcode
+	case "{rsize}":
+		if rr == nil {
+			return EmptyValue
+		}
+		return strconv.Itoa(rr.Len)
+	case "{duration}":
+		if rr == nil {
+			return EmptyValue
+		}
+		return strconv.FormatFloat(time.Since(rr.Start).Seconds(), 'f', -1, 64) + "s"
+	case headerReplacer + "rrflags}":
+		if rr != nil && rr.Msg != nil {
+			return flagsToString(rr.Msg.MsgHdr)
 		}
+		return EmptyValue
 	}
+	return EmptyValue
+}
 
-	// Header placeholders (case-insensitive)
-	rep.replacements[headerReplacer+"id}"] = strconv.Itoa(int(r.Id))
-	rep.replacements[headerReplacer+"opcode}"] = strconv.Itoa(r.Opcode)
-	rep.replacements[headerReplacer+"do}"] = boolToString(req.Do())
-	rep.replacements[headerReplacer+"bufsize}"] = strconv.Itoa(req.Size())
-
-	return rep
+// New makes a new replacer. This only needs to be called once in the setup and then call Replace for each incoming message.
+// A replacer is safe for concurrent use.
+func New() Replacer {
+	return Replacer{
+		valueFunc: value,
+		labels:    labels,
+	}
 }
 
-// Replace performs a replacement of values on s and returns
-// the string with the replaced values.
-func (r replacer) Replace(s string) string {
-
-	// declare a function that replace based on header matching
-	fscanAndReplace := func(s string, header string, replace func(string) string) string {
-		b := strings.Builder{}
-		for strings.Contains(s, header) {
-			idxStart := strings.Index(s, header)
-			endOffset := idxStart + len(header)
-			idxEnd := strings.Index(s[endOffset:], "}")
-			if idxEnd > -1 {
-				placeholder := strings.ToLower(s[idxStart : endOffset+idxEnd+1])
-				replacement := replace(placeholder)
-				if replacement == "" {
-					replacement = r.emptyValue
-				}
-				b.WriteString(s[:idxStart])
-				b.WriteString(replacement)
-				s = s[endOffset+idxEnd+1:]
-			} else {
-				break
-			}
+// Replace performs a replacement of values on s and returns the string with the replaced values.
+func (r Replacer) Replace(ctx context.Context, state request.Request, rr *dnstest.Recorder, s string) string {
+	for _, placeholder := range r.labels {
+		if strings.Contains(s, placeholder) {
+			s = strings.Replace(s, placeholder, r.valueFunc(state, rr, placeholder), -1)
 		}
-		b.WriteString(s)
-		return b.String()
 	}
 
-	// Header replacements - these are case-insensitive, so we can't just use strings.Replace()
-	s = fscanAndReplace(s, headerReplacer, func(placeholder string) string {
-		return r.replacements[placeholder]
-	})
+	// Metadata label replacements. Scan for {/ and search for next }, replace that metadata label with
+	// any meta data that is available.
+	b := strings.Builder{}
+	for strings.Contains(s, labelReplacer) {
+		idxStart := strings.Index(s, labelReplacer)
+		endOffset := idxStart + len(labelReplacer)
+		idxEnd := strings.Index(s[endOffset:], "}")
```

Comment:
Should the "}" be a constant (like "{/" is a constant) ?

Reply:
maybe, don't see the value of making the indirection here.